File: apps/mobile/src/lib/demo-configuration.ts
import LMCommentAndroid, {
  type DemoConfiguration,
  type Tone,
} from '@lm-comment/android';
import Constants from 'expo-constants';
import { Platform } from 'react-native';

export type DemoBootstrapState =
  | 'configured'
  | 'already-configured'
  | 'not-bundled'
  | 'unavailable';

const TONES = new Set<Tone>(['natural', 'professional', 'friendly', 'witty', 'concise']);
let pendingBootstrap: Promise<DemoBootstrapState> | null = null;

export function getBundledDemoConfiguration(): DemoConfiguration | null {
  const value = Constants.expoConfig?.extra?.lmCommentDemo;
  if (!value || typeof value !== 'object') return null;

  const candidate = value as Record<string, unknown>;
  const relayBaseUrl = typeof candidate.relayBaseUrl === 'string'
    ? candidate.relayBaseUrl.trim()
    : '';
  const demoToken = typeof candidate.demoToken === 'string' ? candidate.demoToken.trim() : '';
  const defaultTone = candidate.defaultTone;
  const optionCount = candidate.optionCount;
  const demoMode = candidate.demoMode;
  if (
    !relayBaseUrl
    || demoToken.length < 12
    || typeof defaultTone !== 'string'
    || !TONES.has(defaultTone as Tone)
    || !Number.isInteger(optionCount)
    || (optionCount as number) < 1
    || (optionCount as number) > 3
    || typeof demoMode !== 'boolean'
  ) {
    return null;
  }

  return {
    relayBaseUrl,
    demoToken,
    defaultTone: defaultTone as Tone,
    optionCount: optionCount as 1 | 2 | 3,
    demoMode,
  };
}

async function runBootstrap(): Promise<DemoBootstrapState> {
  if (Platform.OS !== 'android') return 'unavailable';
  const configuration = getBundledDemoConfiguration();
  if (!configuration) return 'not-bundled';

  const status = await LMCommentAndroid.getDemoConfigurationStatus();
  if (status.relayHostname && status.isDemoTokenConfigured) return 'already-configured';

  await LMCommentAndroid.configureDemo(configuration);
  return 'configured';
}

/** Idempotently persists judge defaults without overwriting hidden settings. */
export function bootstrapDemoConfiguration(): Promise<DemoBootstrapState> {
  pendingBootstrap ??= runBootstrap();
  return pendingBootstrap;
}

File: modules/lm-comment-android/android/src/main/java/com/grimnej/lmcomment/config/DemoConfiguration.kt
package com.grimnej.lmcomment.config

enum class Tone(val wireValue: String) {
    NATURAL("natural"),
    PROFESSIONAL("professional"),
    FRIENDLY("friendly"),
    WITTY("witty"),
    CONCISE("concise");

    companion object {
        fun fromWireValue(value: String): Tone =
            entries.firstOrNull { it.wireValue == value }
                ?: throw IllegalArgumentException("Unsupported tone.")
    }
}

data class DemoConfiguration(
    val relayBaseUrl: String,
    val demoToken: String,
    val defaultTone: Tone,
    val optionCount: Int,
    val demoMode: Boolean,
    /** Derived by validation; never persisted or accepted from the JavaScript bridge. */
    val allowDevelopmentHttp: Boolean = false,
)

data class DemoConfigurationStatus(
    val relayHostname: String?,
    val isDemoTokenConfigured: Boolean,
    val defaultTone: Tone,
    val optionCount: Int,
    val demoMode: Boolean,
)
